src/main/scala/mesosphere/marathon/core/storage/migration/MigrationTo1_3_PersistentStore.scala

```diff
@@ -0,0 +1,114 @@
+package mesosphere.marathon.core.storage.migration
+
+// scalastyle:off
+import akka.Done
+import akka.stream.Materializer
+import akka.stream.scaladsl.Source
+import com.typesafe.scalalogging.StrictLogging
+import mesosphere.marathon.core.storage.LegacyStorageConfig
+import mesosphere.marathon.core.storage.repository.{ AppRepository, DeploymentRepository, GroupRepository, Repository, TaskFailureRepository, TaskRepository, VersionedRepository }
+import mesosphere.marathon.metrics.Metrics
+import mesosphere.marathon.state.{ AppDefinition, Group, MarathonTaskState, TaskFailure }
+import mesosphere.marathon.upgrade.DeploymentPlan
+
+import scala.async.Async.{ async, await }
+import scala.concurrent.{ ExecutionContext, Future }
+// scalastyle:on
+
+/**
+  * Migration from Legacy Storage in 1.2 to the new Persistence Storage in 1.3
+  *
+  * Does nothing unless Legacy Storage and New Storage are configured.
+  */
+class MigrationTo1_3_PersistentStore(migration: Migration)(implicit
+  executionContext: ExecutionContext,
+    mat: Materializer,
+    metrics: Metrics) extends StrictLogging {
+  def migrate(): Future[Done] = async {
+    val legacyStore = await(migration.legacyStoreFuture)
+    // This migration only goes from legacy storage -> new storage.
+    (legacyStore, migration.persistenceStore, migration.legacyConfig) match {
+      case (Some(_), Some(_), Some(legacyConfig)) =>
+        val futures = legacyStore.map { _ =>
+          val legacyConfig = {
+            // if persistent store is defined, the config has to be too.
+            require(migration.legacyConfig.isDefined)
+            migration.legacyConfig.get
+          }
+          Seq(
+            migrateTasks(legacyConfig, migration.taskRepo),
+            migrateDeployments(legacyConfig, migration.deploymentRepository),
+            migrateTaskFailures(legacyConfig, migration.taskFailureRepo),
+            // note: we don't actually need to migrate apps (group does it)
+            migrateGroups(legacyConfig, migration.groupRepository)
+          )
+        }.getOrElse(Nil)
+        val summary = await(Future.sequence(futures))
+        logger.info(s"Migrated ${summary.mkString} to 1.3")
+        Done
+      case _ =>
+        logger.info("Skipping Curator Persistence Migration (no legacy store/persistent store in use)")
+        Done
+    }
+  }
+
+  def migrateRepo[T](oldRepo: Repository[_, T], newRepo: Repository[_, T]): Future[Int] = {
+    oldRepo.all().mapAsync(Int.MaxValue) { value =>
```

Comment:
I'm probably jus a burned child, which stumbles across this here over and over again.
To repeat myself: this will put a lot of pressure on zk...